Update/Simplify bot launching CLI params

DanVanAtta opened the pull request:
## Overview
- Add: IDEA headless launcher

- Remove: CLI property `LOBBY_GAME_HOSTED_BY`
- Remove: CLI property `LOBBY_GAME_SUPPORT_EMAIL`
- Remove: `HeadlessGameServerCliParams.java` & `ArgValidationResult.java`, classes did not successfully replace `CliParameters` and are simply redundant at this point.


- Change: `LOBBY_GAME_RECONNECTION` is now defaulted to 2 days (matches prod) instead of accepted as a parameter
- Change: `TRIPLEA_SERVER` is now defaulted to `true` instead of being a parameter
- Change: `LOBBY_GAME_COMMENTS` is now defaulted to `automated_host`
- Change: `GameRunner` will populate `TRIPLEA_NAME` with `messengers.getMessenger().getLocalNode().getName()` to match the value previously used by: `LOBBY_GAME_HOSTED_BY`
- Change: Map folder override if not set is default to users home folder + "/triplea/downloadedMaps"


## Functional Changes
- bot uses fewer parameters to launch

## Manual Testing Performed
- ran local lobby with bot, connected with same player name twice and started a game between the two players.

## Additional Review Notes
- motivation was to see if bot could benefit from env variables to be consistent with other apps. This brought me back to some long-standing issues where launching headless is overly cumbersome.

RoiEXLab commented:
Oh and @prastle should probably know about this

DanVanAtta commented on game-core/src/main/java/org/triplea/game/server/HeadlessGameServer.java:
```diff
@@ -617,79 +594,46 @@ private static void usage() {
-              + " must start with \"Bot\" and be at least 7 characters long and be the same.");
+    String playerName = System.getProperty(TRIPLEA_NAME, "");
+    if (playerName.length() < 7) {
+      log.warning("Invalid argument: " + TRIPLEA_NAME + " must at least 4 characters long");
```
The thinking was now that 'bot' is prepended before this check, to a user they only need to supply 4.  OTOH that is not always the case, EG: `Bot_`, I agree this is under-engineered.

DanVanAtta replied:
Decided to go back to the old behavior for-now:  6914502184fce02350c59ba01b9ec592f5e1fc84

DanVanAtta commented on game-core/src/main/java/org/triplea/game/server/HeadlessGameServer.java:
```diff
@@ -50,16 +45,14 @@
 public class HeadlessGameServer {
-  private static final int LOBBY_RECONNECTION_REFRESH_SECONDS_MINIMUM = 21600;
-  private static final int LOBBY_RECONNECTION_REFRESH_SECONDS_DEFAULT = 2 * LOBBY_RECONNECTION_REFRESH_SECONDS_MINIMUM;
+  private static final int LOBBY_RECONNECTION_REFRESH_SECONDS_DEFAULT = 2 * 21600;
```
What, you're not pleased to know this is SECONDS? :grin:  What else could you possibly want to know in 43 characters?

yeah, these do not mean that much to me either and is not necessarily what I was trying to change. Clarifying/fixing up every line that is touched leads to unreviewable and (in this code-base) very risky updates. I could very easily imagine the criticism of mixing scopes and hard-to-review-diff  had such a refactor been included.  My point is that even if this were a thing I would have wanted to update, I would have left it out for the sake of CR diff, it then coming up in a review is a problem as such an update should have been left out to keep the diff small. The problem is that this creates a gray area of what is related, it is different depending on reviewer. The lack of consistency then makes a damned-if-do, damned-if-you-don't type of situation.

DanVanAtta replied:
This did need fixing, done:   3f0139fd7

ssoloff commented on game-core/src/main/java/org/triplea/game/server/HeadlessGameServer.java:
```diff
@@ -607,79 +586,41 @@ private static void usage() {
-          + " seconds, and should normally be either " + LOBBY_RECONNECTION_REFRESH_SECONDS_DEFAULT + " or "
-          + (2 * LOBBY_RECONNECTION_REFRESH_SECONDS_DEFAULT) + " seconds.");
+    if (Integer.valueOf(System.getProperty(LOBBY_PORT, "0")) <= 0) {
+      log.warning("Invalid argument: " + LOBBY_PORT + " must be numeric");
```
```suggestion
      log.warning("Invalid argument: " + LOBBY_PORT + " must be greater than zero");
```

If you want to check for non-numeric values, you're going to have to catch `NumberFormatException`.

ssoloff replied:
Bumping

ssoloff commented on game-core/src/main/java/games/strategy/engine/framework/GameRunner.java:
```diff
@@ -273,7 +272,6 @@ public static void hostGame(final int port, final String playerName, final Strin
     commands.add("-D" + LOBBY_PORT + "="
         + messengers.getMessenger().getRemoteServerSocketAddress().getPort());
     commands.add("-D" + LOBBY_GAME_COMMENTS + "=" + comments);
-    commands.add("-D" + LOBBY_GAME_HOSTED_BY + "=" + messengers.getMessenger().getLocalNode().getName());
```
Are you sure that `playerName` (used to initialize `TRIPLEA_NAME`) is identical to `messengers.getMessenger().getLocalNode().getName()` in all cases?  There's some weirdness that can go on when you have the same user logged in multiple times and an ordinal gets appended to the node name, but not necessarily the player name (or maybe that's vice versa).

DanVanAtta replied:
Can use `messengers.getMessenger().getLocalNode().getName()` instead: 3cf1e47dc

Given the 'hosted-by' and 'triplea-name' are asserted to be the same by the bot code, and the two values are then used interchangeably by bot code, that if there is any difference then we have a latent bug. I'm much rather for us to simplify and make any such problems more obvious and detectable, particularly when releases will be internal only for some time to come.

DanVanAtta replied:
Testing looked good. I'm reasonably confident on this one but it's hard to be 100% sure.

ssoloff commented on game-core/src/main/java/games/strategy/engine/framework/startup/ui/InGameLobbyWatcher.java:
```diff
@@ -101,7 +101,7 @@ public static InGameLobbyWatcher newInGameLobbyWatcher(
     final @Nullable String host = getLobbySystemProperty(LOBBY_HOST);
     final @Nullable String port = getLobbySystemProperty(LOBBY_PORT);
-    final @Nullable String hostedBy = getLobbySystemProperty(LOBBY_GAME_HOSTED_BY);
+    final @Nullable String hostedBy = getLobbySystemProperty(TRIPLEA_NAME);
```
See warning elsewhere about node names with ordinals.  This is assumed to be a node name.  If it doesn't match the player's node name, the associated lobby watcher may not work correctly.

ssoloff replied:
Hard to tell by inspection.  I'd try running a local lobby with one of the users logged in multiple times.  Have one of that user's nodes exercise some of the code in this class (e.g. watching [not playing] an in-progress game) and see if anything strange happens.

DanVanAtta replied:
testing looked good! thanks for the suggestion @ssoloff

ssoloff commented on game-core/src/main/java/org/triplea/game/server/HeadlessGameServer.java:
```diff
@@ -607,79 +586,41 @@ private static void usage() {
-      printUsage = true;
+    if (!ClientSetting.mapFolderOverride.isSet()) {
+      ClientSetting.mapFolderOverride.setValue(
+          Paths.get(SystemProperties.getUserHome(), "/triplea/downloadedMaps"));
```
At a minimum, I'd recommend using `ClientFileSystemHelper#getUserRootFolder()` (see how `ClientSetting#userMapsFolderPath` is initialized).  You could probably just use `ClientSetting#userMapsFolderPath` itself.

Setting `ClientSetting#mapFolderOverride` also seems to duplicate the code in `ArgParser`.  Is it possible to have `ArgParser` set this default value if it doesn't encounter `MAP_FOLDER`?

DanVanAtta replied:
I rolled back the default value of maps folder (for now; it was rolled back to reduce complexity of this branch, leaving for a follow-up task)

ssoloff replied:
> Correction, went ahead with the defaulting code per the suggestion to use `ClientFileSystemHelper`.

I don't see a commit that implemented this.  Did you reconsider, or am I misunderstanding your intent?

DanVanAtta replied:
Thanks for the follow-up, I did not get this done completely. I pushed:  adb6d0568 just now to get the path from `ClientFileSystemHelper`.

ssoloff commented on game-core/src/main/java/org/triplea/game/server/HeadlessGameServer.java:
```diff
@@ -607,79 +586,41 @@ private static void usage() {
-      log.warning(
-          "Invalid argument: " + LOBBY_GAME_COMMENTS + " must contain the string \"automated_host\".");
+    if (Integer.valueOf(System.getProperty(TRIPLEA_PORT, "0")) <= 0) {
+      log.warning("Invalid argument: " + TRIPLEA_PORT + " must be numeric");
```
```suggestion
      log.warning("Invalid argument: " + TRIPLEA_PORT + " must be greater than zero");
```

If you want to check for non-numeric values, you're going to have to catch `NumberFormatException`.

DanVanAtta replied:
Fixed in: 198abcf57

ssoloff commented on game-core/src/main/java/org/triplea/game/server/HeadlessGameServer.java:
```diff
@@ -580,19 +567,11 @@ public Chat getChat() {
-    }
-
     ClientSetting.initialize();
+    System.setProperty(LOBBY_GAME_COMMENTS, "automated_host");
```
Is the intention here to lock down these properties or just to provide defaults?  If the former, these calls probably need to be after `ArgParser.handleCommandLineArgs(args)` to prevent a user from specifying different values on the command line.

DanVanAtta replied:
Motivation is:
1. Simplify/reduce, best thing I can think to do before seriously looking at environment variables for secrets or other configuration
2. Make it easier to run bots

Locking down is only going to be so successful given the hosting user of a bot has access to the source code. Fail-early still applies though for bad/missing parameters of course, that probably should be the goal of the validation done here.

DanVanAtta commented on game-headless/scripts/run_bot:
```diff
@@ -78,17 +63,6 @@ readonly LOBBY_HOST=lobby.triplea-game.org
-# The default value is meaningless.  IT IS STRONGLY RECOMMENDED THAT YOU
-# CHANGE THIS VARIABLE TO A VALID EMAIL ADDRESS THAT YOU CONTROL.
-#
-readonly LOBBY_SUPPORT_EMAIL="${USER}@localhost.localdomain"
```
Intentional to remove this one, YAGNI; this has never been used.

This is maybe getting controversial, IMO the main audience of bots is 'us', any idea that we would get long term bot hosting support from someone else has not panned out. Anyone seriously hosting bots will contact us; Anyone with bots acting up we will deal with by booting and notifying in most likely gitter, if malicious bots then the email will be a dead-end anyways.

DanVanAtta replied:
To be clear, Bung was the system admin that took over from SgBridges, who owned that responsibility while Veqryn focused on the code side. Bung handed system admin to me. @prastle is as far as I'm concerned part of the TripleA maintainers group. If he had trouble running official TripleA bots, it was a failure on my part, lack of communication between him and I, and a lack of clarity that I am responsible for lobby and bot uptime and deployments.

I'm pretty sure though that @prastle was helping others get bots running and was entertaining the idea of having others help official support bots. IIRC that was before we doubled the available bots and no longer have that same pressure. Regardless, there are still reasons for having run-scripts that are known to be working, that is mutually exclusive to the question of if we want to have a given parameter in that script. IE: running the bot at all is a different concern from how to password it, once it is running.


@prastle as far as I'm concerned is part of the TripleA maintainers group. His role as the leader of moderators and his support for running the server are really valuable, even if he is not adding code.

I view interactions with maintainers compared to other users as fundamentally different. To help prastle run a script, we need to simple work with him and make sure maintainers can run scripts. That is why infrastructure was created. 

I'm not sure if you are big fan of Locke, but I found it's interesting relative to open-source: https://plato.stanford.edu/entries/locke-political/#Pro

We trace BDFL to @ron-murhammer for game-code. System-admin traces to me. I'll not '@' everone as that is annoying, but happy to do the official ownership tracing back to the creator of who first ran servers.

sbridges was first running the servers, veqryn was focused on the code and delegated system-admin work for the lifetime of the project. It became a problem that the two groups became so distinct.
Then Bung ran the servers for a long time with help from Frankie and American Knight. Eventaully Bung retired as did all other system admins, there was a gap and eventually Bung handed system admin to me. So in that perverse way, @prastle when running the official bots, is doing so as part of official TripleA.

DanVanAtta replied:
I'm really confident on this feature, that even if it were ever used we would still not need or want it, or use it. 

Having reference scripts to launch bots is a different concern compared to having a param to specify your email address.  "having to figure out what's wrong with their script" is not related to this parameter existing or not.

DanVanAtta commented on game-headless/scripts/run_bot:
```diff
@@ -46,21 +46,6 @@ readonly BOT_PORT=3300
-# that case, set the password to a non-empty string and communicate it to
-# your friends in some manner (e.g. via email).
-#
-readonly BOT_PASSWORD=
```
Bots are hosted for other people, password protecting bots does not really make sense. We likely do not need to pass password as a parameter for bots.

We have a lot of time before next release. IMO we need to take YAGNI pretty seriously and favor removing unused features

ssoloff replied:
There were some queries in 2018 of users wanting to host headless bots for the use of _their_ friends.  I'm pretty sure their intention was to restrict access so an arbitrary lobby user couldn't simply start a game on their private bot.

DanVanAtta replied:
Was that before we doubled the size of the bot fleet? 

(this is a moot consideration as I rolled back the changes in the script  files and removed just the parameters that are now hardcoded). If it were up to me we would YAGNI the :hankey: out of bots and headless and likely have something pretty different by the time we hit next release. The fact that we had required/non-required bugs, two different usage texts that could kick in is a pretty big failing. I'm a fan of not blaming people, that only leaves the blame on code complexity. Having features that are not strictly asked for or needed is not a good thing.  I appreciate the example, but even then, one example over the last *year*, I can't think of any before that, and even in the example a password is not strictly needed.

There is a connection lag on late-stage games when new players join. I'm not sure if that is fixed. Largely that is the reason for passwording, people are eager to check out the round 24 game, with a 20 second game-freeze on each new person joining, it quickly adds minutes to game-turns. IMO we should be getting it so that turns in TripleA are minutes, preferably *a* minute. That goal is why I started dev'ing for TripleA to begin with, to make it faster to play a single turn, let alone the entire game. There's been a lot to do...

DanVanAtta commented on game-headless/scripts/run_bot:
```diff
@@ -114,15 +88,9 @@ java \
-    -Ptriplea.lobby.game.supportPassword="$LOBBY_SUPPORT_PASSWORD" \
     -Ptriplea.lobby.host="$LOBBY_HOST" \
     -Ptriplea.lobby.port="$LOBBY_PORT" \
+    -Ptriplea.lobby.game.supportPassword="$LOBBY_SUPPORT_PASSWORD" \
```
I rolled back the changes in the script files and then re-applied, removing  just the parameters that are now hardcoded

DanVanAtta commented on game-headless/scripts/run_bot.bat:
```diff
@@ -78,17 +63,6 @@ REM Under normal circumstances, you will not have to change this variable.
-REM The default value is meaningless.  IT IS STRONGLY RECOMMENDED THAT YOU
-REM CHANGE THIS VARIABLE TO A VALID EMAIL ADDRESS THAT YOU CONTROL.
-REM
-SET LOBBY_SUPPORT_EMAIL=%USERNAME%@localhost.localdomain
```
Resolved, restored in:  d054b783b5860e995f3b26183e9638db5739924a

DanVanAtta commented on game-headless/scripts/run_bot.bat:
```diff
@@ -46,21 +46,6 @@ REM you will not have to change this variable.
-REM that case, set the password to a non-empty string and communicate it to
-REM your friends in some manner (e.g. via email).
-REM
-SET BOT_PASSWORD=
```
Resolved, restored in: d054b78

RoiEXLab commented on game-core/src/main/java/org/triplea/game/server/HeadlessGameServer.java:
```diff
@@ -36,28 +34,25 @@
 public class HeadlessGameServer {
-  private static final int LOBBY_RECONNECTION_REFRESH_SECONDS_MINIMUM = 21600;
-  private static final int LOBBY_RECONNECTION_REFRESH_SECONDS_DEFAULT = 2 * LOBBY_RECONNECTION_REFRESH_SECONDS_MINIMUM;
+  private static final int LOBBY_RECONNECTION_REFRESH_SECONDS_DEFAULT = (int) TimeUnit.HOURS.toSeconds(12);
```
For future reference: 12 hours is definitely too long

DanVanAtta replied:
I was thinking the same, let's follow up in: https://github.com/triplea-game/triplea/issues/4552  "lobby.game.reconnection time does not seem right"

prastle commented:
Can someone restart the 400 series and 500 series

Jan 12 00:28:56 bot45_atlanta_ga run_bot.sh[23885]: Jan 12, 2019 12:28:56 AM games.strategy.engine.data.GameParser parseGameLoader
Jan 12 00:28:56 bot45_atlanta_ga run_bot.sh[23885]: INFO: Loader tag is being ignored
Jan 12 00:28:56 bot45_atlanta_ga run_bot.sh[23885]: Jan 12, 2019 12:28:56 AM games.strategy.engine.data.GameParser parseGameLoader
Jan 12 00:28:56 bot45_atlanta_ga run_bot.sh[23885]: INFO: Loader tag is being ignored
Jan 12 00:28:56 bot45_atlanta_ga run_bot.sh[23885]: Jan 12, 2019 12:28:56 AM games.strategy.engine.data.GameParser parseGameLoader
Jan 12 00:28:56 bot45_atlanta_ga run_bot.sh[23885]: INFO: Loader tag is being ignored
Jan 12 00:28:58 bot45_atlanta_ga run_bot.sh[23885]: Jan 12, 2019 12:28:58 AM games.strategy.engine.data.GameParser parseGameLoader
Jan 12 00:28:58 bot45_atlanta_ga run_bot.sh[23885]: INFO: Loader tag is being ignored
Jan 12 00:29:04 bot45_atlanta_ga run_bot.sh[23885]: Jan 12, 2019 12:29:04 AM games.strategy.engine.data.GameParser parseGameLoader
Jan 12 00:29:04 bot45_atlanta_ga run_bot.sh[23885]: INFO: Loader tag is being ignored
~

prastle commented:
Never Mind they finally launched 14 min

RoiEXLab commented on game-core/src/main/java/org/triplea/game/server/HeadlessGameServer.java:
```diff
@@ -109,21 +104,13 @@ private HeadlessGameServer() {
         log.log(Level.WARNING, "Failed to restart Lobby watcher", e);
       }
-    }, reconnect, reconnect, TimeUnit.SECONDS);
+    }, LOBBY_RECONNECTION_REFRESH_SECONDS_DEFAULT, LOBBY_RECONNECTION_REFRESH_SECONDS_DEFAULT, TimeUnit.SECONDS);
```
Just a minor thing I noticed:
Instead of storing the amount of seconds within 12 hours, it would make more sense to store the 12 hours and use `TimeUnit.HOURS` in this line instead.

DanVanAtta replied:
I like the idea; I'm reluctant to change the scope of this branch. Did we find out if 12+ hours is correct?

DanVanAtta replied:
Excellent context @ssoloff , thank you. Did you run across per-chance the reason why the disconnect was added?

DanVanAtta replied:
Following up, in prod we set this value to 2 days. I've updated this branch to match: c45d87517

I've opened issue: https://github.com/triplea-game/triplea/issues/4552 to discuss what we would like to do long-term with this property and if it is correct or not: "lobby.game.reconnection time does not seem right"

ssoloff replied:
> Did you run across per-chance the reason why the disconnect was added?

Unfortunately, no (or, at least, I didn't recognize it as such if I did come across it).

DanVanAtta commented:
Some updates to report:
1. tested the `triplea.name` vs `lobby.hosted.by` settings on local with two players with the same name, looked reasonable and could start a game and chat between them.
2. removed entire bot support email feature to match intent instead of just removing the CLI parameter.
3. re-did same testing from (1)
4. Updated overview of this PR to have a clean 'add/remove/change' list

ssoloff commented on game-core/src/main/java/games/strategy/engine/lobby/client/ui/LobbyGamePanel.java:
```diff
@@ -200,7 +200,6 @@ private void mouseOnGamesList(final MouseEvent e) {
 
   private Collection<Action> getBotAdminGamesListContextActions(final GameDescription description) {
```
`description` is now an unused parameter.

DanVanAtta replied:
Fixed in: ddbe1cb1b

ssoloff commented on game-core/src/main/java/games/strategy/engine/framework/startup/ui/RemoteHostUtils.java:
```diff
@@ -110,7 +110,7 @@ public String shutDownHeadlessHostBot(final String hashedPassword, final String
     if (instance == null) {
```
Not suggesting this be done in this PR--because it would probably require additional testing--but now that `remoteShutdown()` and `getSalt()` are static, these `instance == null` checks seem to no longer be necessary.  They appear to have only been added to prevent an NPE.  However, if they are truly required, the checks should be hidden in the static methods themselves.

ssoloff commented on game-core/src/main/java/games/strategy/engine/lobby/server/GameDescription.java:
```diff
@@ -256,7 +251,6 @@ public void readExternal(final ObjectInput in) throws IOException {
     passworded = in.readBoolean();
     engineVersion = in.readUTF();
     gameVersion = in.readUTF();
-    botSupportEmail = in.readUTF();
```
Since you're writing out an empty string for this field in `writeExternal()` (presumably to preserve lobby compatibility), do you also need to read (and discard) a string here for the same reason?  Otherwise the stream may become unsynchronized if multiple `GameDescription` instances have been serialized.

ssoloff commented on game-core/src/test/java/org/triplea/lobby/common/GameDescriptionTest.java:
```diff
@@ -3,46 +3,49 @@
   }
 
-  @Nested
-  final class IsBotTest {
```
FYI, nesting tests like this has the side effect of de-triggering the Eclipse static analyzer from wanting to say the test methods can be static (which they can't due to a JUnit requirement); that seemed preferable to adding a `@SuppressWarnings("static-method")` in such cases.  It also has the benefit of not repeating the name of the method under test in each test case.

DanVanAtta replied:
I should have called this one out:
- `@MethodSource` methods must be static
- nested classes may not contain nested methods

That was the motivation for removing the nesting. On a side note, IDEA declares the nested test classes as 'emtpy-classes' and we likely should suppress that warning. I agree to the benefit of not repeating test names.

ssoloff replied:
> nested classes may not contain nested methods

Ah, yes, _that problem_ again. :smile:

Here's a slightly different idea you can consider if it is acceptable:

Since you're already forced to host the argument factory in the top-level class, you could switch to using `@ArgumentsSource` instead of `@MethodSource`.  This has the following benefits:

* It allows you to move the test methods back inside the nested fixture.
* It removes the need to suppress the unused method warning because the factory is now referenced statically rather than via reflection.

But it has the following drawbacks:

* The factory is no longer in the same scope as the test methods.
* It's a few more lines of code for the class declaration beyond just the method definition.

ssoloff replied:
That's fine as it's consistent with most of our data-driven tests.  The addition of `@ParameterizedTest` recently was to overcome the smell of a Mockito `reset()`, which isn't present here.  So, old-school list is good. :+1:

ssoloff commented on game-core/src/test/java/org/triplea/lobby/common/GameDescriptionTest.java:
```diff
@@ -3,46 +3,49 @@
 
-      assertThat(gameDescription.isBot(), is(false));
-    }
+  private static Collection<GameDescription> notBotDescriptions() {
```
Gonna need a `@SuppressWarnings("unused")` here due to the reflective invocation.  (I upgraded to Eclipse 2018-12, and it still generates this warning.)

DanVanAtta replied:
Per: https://github.com/triplea-game/triplea/pull/4545#discussion_r248130230
Rolled back update that added this method: f452e68

ssoloff commented on game-core/src/main/java/games/strategy/engine/framework/GameRunner.java:
```diff
@@ -268,13 +267,12 @@ public static void hostGame(final int port, final String playerName, final Strin
     ProcessRunnerUtil.populateBasicJavaArgs(commands);
     commands.add("-D" + TRIPLEA_SERVER + "=true");
     commands.add("-D" + TRIPLEA_PORT + "=" + port);
-    commands.add("-D" + TRIPLEA_NAME + "=" + playerName);
```
`playerName` is now an unused parameter to this method.
